src/providers/naver.ts

```diff
@@ -0,0 +1,51 @@
+import { createOAuth2Request, sendTokenRequest } from "../request.js";
+
+import type { OAuth2Tokens } from "../oauth2.js";
+
+const authorizationEndpoint = "https://nid.naver.com/oauth2.0/authorize";
+const tokenEndpoint = "https://nid.naver.com/oauth2.0/token";
+
+export class Naver {
+	private clientId: string;
+	private clientSecret: string;
+	private redirectURI: string;
+
+	constructor(clientId: string, clientSecret: string, redirectURI: string) {
+		this.clientId = clientId;
+		this.clientSecret = clientSecret;
+		this.redirectURI = redirectURI;
+	}
+
+	public createAuthorizationURL(state: string): URL {
+		const url = new URL(authorizationEndpoint);
+		url.searchParams.set("response_type", "code");
+		url.searchParams.set("client_id", this.clientId);
+		url.searchParams.set("redirect_uri", this.redirectURI);
+		url.searchParams.set("state", state);
+		return url;
+	}
+
+	public async validateAuthorizationCode(state: string, code: string): Promise<OAuth2Tokens> {
+		const body = new URLSearchParams();
+		body.set("grant_type", "authorization_code");
+		body.set("client_id", this.clientId);
+		body.set("client_secret", this.clientSecret);
+		body.set("code", code);
+		body.set("state", state);
```

Comment:
Is `state` required? If it is, it's not compliant with the spec and I can't merge it. Looks like Auth.js doesn't pass it anyway tho?

https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/providers/naver.ts

Reply:
According to the official document, `state` is a required property.

> https://developers.naver.com/docs/login/devguide/devguide.md#3-4-2-네이버-로그인-연동-url-생성하기

<img width="900" alt="Screenshot 2024-10-30 at 10 23 22 AM" src="https://github.com/user-attachments/assets/eeb4c05d-78df-4876-985b-b74601ec2721">

I'm not sure why the code in nextauth doesn't specify `state`, but it also works 🤔🤔

Reply:
Looking at other example code, I don't think `state` is required. Do you have a working implementation locally? Would be great if you could test if `state` is required or not

Reply:
I've actually used these patches in my personal project, allowing me to easily test whether `state` is required.

As a result, it appears that `state` is not required, and it seems the provider does not validate it 😢 

I need to add a new commit to remove `state` 👀 

### Test case 1

It works when `state` is included in both `createAuthorizationURL` and `validateAuthorizationCode`.

<details>
<summary>case 1</summary>
<p>

```ts
  public createAuthorizationURL(state: string): URL {
    const url = new URL(authorizationEndpoint);

    url.searchParams.set("response_type", "code");
    url.searchParams.set("client_id", this.clientId);
    url.searchParams.set("redirect_uri", this.redirectURI);
    url.searchParams.set("state", state); ///// HERE

    return url;
  }

  public async validateAuthorizationCode(
    state: string,
    code: string,
  ): Promise<OAuth2Tokens> {
    const body = new URLSearchParams();

    body.set("grant_type", "authorization_code");
    body.set("client_id", this.clientId);
    body.set("client_secret", this.clientSecret);
    body.set("code", code);
    body.set("state", state); ///// HERE
    body.set("redirect_uri", this.redirectURI);

    const request = createOAuth2Request(tokenEndpoint, body);
    const tokens = await sendTokenRequest(request);

    return tokens;
  }
```

</p>
</details>

### Test case 2

It works when `state` is included only in `createAuthorizationURL`.

<details>
<summary>case 2</summary>
<p>

```ts
  public createAuthorizationURL(state: string): URL {
    const url = new URL(authorizationEndpoint);

    url.searchParams.set("response_type", "code");
    url.searchParams.set("client_id", this.clientId);
    url.searchParams.set("redirect_uri", this.redirectURI);
    url.searchParams.set("state", state); ///// HERE

    return url;
  }

  public async validateAuthorizationCode(
    state: string,
    code: string,
  ): Promise<OAuth2Tokens> {
    const body = new URLSearchParams();

    body.set("grant_type", "authorization_code");
    body.set("client_id", this.clientId);
    body.set("client_secret", this.clientSecret);
    body.set("code", code);
    body.set("redirect_uri", this.redirectURI);

    const request = createOAuth2Request(tokenEndpoint, body);
    const tokens = await sendTokenRequest(request);

    return tokens;
  }
```

<p>
</details>

### Test case 3

It works even when `state` is not included anywhere

<details>
<summary>case 3</summary>
<p>

```ts
  public createAuthorizationURL(state: string): URL {
    const url = new URL(authorizationEndpoint);

    url.searchParams.set("response_type", "code");
    url.searchParams.set("client_id", this.clientId);
    url.searchParams.set("redirect_uri", this.redirectURI);

    return url;
  }

  public async validateAuthorizationCode(
    state: string,
    code: string,
  ): Promise<OAuth2Tokens> {
    const body = new URLSearchParams();

    body.set("grant_type", "authorization_code");
    body.set("client_id", this.clientId);
    body.set("client_secret", this.clientSecret);
    body.set("code", code);
    body.set("redirect_uri", this.redirectURI);

    const request = createOAuth2Request(tokenEndpoint, body);
    const tokens = await sendTokenRequest(request);

    return tokens;
  }
```

<p>
</details>

Reply:
great, thanks!

Reply:
yeah thank you for your review ! 

I've fixed it here: 92be851288b176263a5b230f8ccf857d026e6d19